plugins/woocommerce/includes/class-wc-session-handler.php

```diff
@@ -627,10 +627,23 @@ public function maybe_update_nonce_user_logged_out( $uid, $action ) {
 	public function cleanup_sessions() {
 		global $wpdb;
 
-		$wpdb->query( $wpdb->prepare( 'DELETE FROM %i WHERE session_expiry < %d', $this->_table, time() ) );
-
-		if ( class_exists( 'WC_Cache_Helper' ) ) {
-			WC_Cache_Helper::invalidate_cache_group( WC_SESSION_CACHE_GROUP );
+		// The initial query can be slow for big session tables, but it doesn't make sense to introduce an index covering session_expiry due to high cardinality.
+		$expired_sessions = $wpdb->get_results( $wpdb->prepare( 'SELECT session_id, session_key FROM %i WHERE session_expiry < %d', $this->_table, time() ) );
+		if ( null !== $expired_sessions ) {
+			$batch_size   = 100;
```

Comment:
Batching could work here, though I wonder if we could just batch the delete process instead of the select then delete (assuming I'm correct on the exclusive lock).

Depending on the performance issues we're able to replicate, we might want to also async batch these, but I think I'm missing some context around the performance issues for this issue.

Reply:
> we might want to also async batch these

It is already an async context (sessions cleanup is scheduled and executed with a 12-hour interval).